tests/cpp_tests/testutils.cpp

```diff
@@ -0,0 +1,369 @@
+    }
+    if (groups) {
+      weights->reserve(nrows);
+    }
+
+    int32_t group = 0;
+
+    for (int32_t row = 0; row < nrows; row++) {
+      labels->push_back(rand.NextFloat());
+      if (weights) {
+        weights->push_back(rand.NextFloat());
+      }
+      if (init_scores) {
+        for (int32_t i = 0; i < nclasses; i++) {
+          init_scores->push_back(rand.NextFloat());
+        }
+      }
+      if (groups) {
+        if (rand.NextFloat() > 0.95) {
+          group++;
+        }
+        groups->push_back(group);
+      }
+    }
+  }
+
+  void TestUtils::StreamDenseDataset(DatasetHandle dataset_handle,
+    int32_t nrows,
+    int32_t ncols,
+    int32_t nclasses,
+    int32_t batch_count,
+    const std::vector<double>* features,
+    const std::vector<float>* labels,
+    const std::vector<float>* weights,
+    const std::vector<double>* init_scores,
+    const std::vector<int32_t>* groups) {
+    int result = LGBM_DatasetSetWaitForManualFinish(dataset_handle, 1);
+    EXPECT_EQ(0, result) << "LGBM_DataseSetWaitForManualFinish result code: " << result;
+
+    if ((nrows % batch_count) != 0) {
+      Log::Fatal("This utility method only handles nrows that are a multiple of batch_count");
+    }
+
+    const double* features_ptr = features->data();
+    const float* labels_ptr = labels->data();
+    const float* weights_ptr = nullptr;
+    if (weights) {
+      weights_ptr = weights->data();
+    }
+
+    // Since init_scores are in a column format, but need to be pushed as rows, we have to extract each batch
+    std::vector<double> init_score_batch;
+    const double* init_scores_ptr = nullptr;
+    if (init_scores) {
+      init_score_batch.reserve(nclasses * batch_count);
+      init_scores_ptr = init_score_batch.data();
+    }
+
+    const int32_t* groups_ptr = nullptr;
+    if (groups) {
+      groups_ptr = groups->data();
+    }
+
+    auto start_time = std::chrono::steady_clock::now();
+
+    for (int32_t i = 0; i < nrows; i += batch_count) {
+      if (init_scores) {
+        init_scores_ptr = CreateInitScoreBatch(&init_score_batch, i, nrows, nclasses, batch_count, init_scores);
+      }
+
+      result = LGBM_DatasetPushRowsWithMetadata(dataset_handle,
+                                                features_ptr,
+                                                1,
+                                                batch_count,
+                                                ncols,
+                                                i,
+                                                labels_ptr,
+                                                weights_ptr,
+                                                init_scores_ptr,
+                                                groups_ptr,
+                                                0);
+      EXPECT_EQ(0, result) << "LGBM_DatasetPushRowsWithMetadata result code: " << result;
+      if (result != 0) {
+        FAIL() << "LGBM_DatasetPushRowsWithMetadata failed";
+      }
```

Comment:
Looks like a duplicated code...
```suggestion
      EXPECT_EQ(0, result) << "LGBM_DatasetPushRowsWithMetadata result code: " << result;
```

Reply:
Oddly enough, "EXPECT_EQ" does not terminate the test right away, so I added the FAIL on purpose (because otherwise it prints out a ton of messages before failing the test).  Will add a comment